Make ignored attributes an array

davidvgalbraith opened the pull request:
Currently you initialize Bubo's ignored attributes as an object whose keys are the attributes you want to ignore. This is weird. It should just be an array of the attributes you want to ignore. And then the `ignored_attributes` all over the Bubo code are `std::vector`s instead of v8 objects. This didn't noticeably change performance.

@demmer

Commit:
Make ignored attributes an array

davidvgalbraith commented on src/bubo.cc:
```diff
@@ -55,8 +55,19 @@ NAN_METHOD(Bubo::Initialize)
-    Local<Value> ignoredVal = Nan::Get(opts, ignoredAttributes).ToLocalChecked();
-    Local<Object> ignored = Nan::To<Object>(ignoredVal).ToLocalChecked();
+    Local<Value> ignored_value = Nan::Get(opts, ignoredAttributes).ToLocalChecked();
+    Local<Object> ignored_array = Nan::To<Object>(ignored_value).ToLocalChecked();
```
This is weird but Nan doesn't have `To<Array>` for some reason and I didn't see how to do it. (https://github.com/nodejs/nan/blob/master/doc/converters.md#api_nan_to)

demmer replied:
I've noticed that before and I have no idea why.

demmer replied:
Actually it's because v8::Value doesn't actually expose a ToArray function for some reason: http://izs.me/v8-docs/classv8_1_1Value.html.

Odd because there is an IsArray function...

demmer commented on src/attrs-table.cc:
```diff
@@ -88,11 +92,13 @@ bool AttributesTable::prepare_entry_buffer(const v8::Local<v8::Object>& pt,
+        v8::String::Utf8Value tag(key);
+        std::string tag_str(*tag);
+
+        if (!ignored_attributes_.empty() && _contains(ignored_attributes_, tag_str)) {
```
The previous `IsEmpty()` check was just verifying that ignored_attributes_  was actually set to a valid object. That's not possible any more, and iterating over an empty vector is a safe operation, so there's no need for the explicit `empty()` check.

demmer replied:
Oh wait - scratch that. In this new approach you're actually copying the ignored_attributes vector instead of sharing it via a pointer.

So in fact, you should change `ignored_attributes` to be a `std::vector<std::string>*` and then make sure it's non-null before calling `_contains`.

demmer commented on src/bubo-cache.cc:
```diff
@@ -23,8 +23,8 @@ BuboCache::~BuboCache() {
 
-void BuboCache::initialize(v8::Local<v8::Object> ignoredAttrs) {
-    ignored_attributes_.Reset(ignoredAttrs);
+void BuboCache::initialize(std::vector<std::string> ignoredAttrs) {
```
In general I've always followed the convention of using const references when passing non-trivial types by value since it avoids having to copy the data structure twice. Obviously the performance impact is minimal here and it's possible that newer C++ compilers are smarter anyway, but still changing this argument to `const std::vector<std::string>& ignoredAttrs` would (I think) result in the same result and a more consistent style.

davidvgalbraith replied:
I always do whatever has the least ampersands, but that's just because I'm a bewildered chicken. I'll put some ampersands in.

Commit:
Make attrs-table.ignored_attributes_ a pointer

davidvgalbraith commented:
I converted it to ampersands.

demmer committed:
verify that ignoredAttributes is in fact an array

It turns out it is possible to get an Array out of an Object, so
let's do that and use a cleaner iteration model with a for loop
instead.

demmer commented:
Just pushed a commit that does extract an Array out of the reference and added a test to make sure it properly fails if the argument is not in fact an array.

Assuming that looks good to you, I'm +1 for the changes you just did and this can go in.

Commit:
Prefix MaybeLocal with Nan so it builds
